### PR TITLE
fix: bump dill to >= 0.3.6, prevents tests hanging with python3.11

### LINKED ISSUE #7916: pylint/2.15.8: python/3.11.1: debian: tests/test_check_parallel.py tests fail/hang indefinitely
### Bug description

Running tests with python3.11, makes some tests/test_check_parallel.py either fail or hang indefinitely:

```
I: pybuild base:240: cd /build/pylint-2.15.8/.pybuild/cpython3_3.11/build; python3.11 -m pytest -vvvv -k 'not test_pkginfo and not test_do_not_import_files_from_local_directory and not test_import_plugin_from_local_directory_if_pythonpath_cwd and not test_can_list_directo
ries_without_dunder_init and not test_fail_on_exit_code and not test__test_environ_pythonpath_no_arg'
============================= test session starts ==============================
platform linux -- Python 3.11.1, pytest-7.1.2, pluggy-1.0.0+repack -- /usr/bin/python3.11
cachedir: .pytest_cache
benchmark: 3.2.2 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /build/pylint-2.15.8, configfile: setup.cfg
plugins: benchmark-3.2.2
collecting ... collected 2057 items / 15 deselected / 2042 selected

tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_initialize FAILED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_initialize_pickling PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_check_single_file_uninitialised PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_check_single_file_no_checkers FAILED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_linter_with_unpickleable_plugins_is_pickleable PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_check_sequential_checker FAILED [  0%]
tests/test_check_parallel.py::TestCheckParallel::test_sequential_checkers_work Terminated
```

the last had to be `kill`ed

the same code, ran with python 3.10 works fine:

```
I: pybuild base:240: cd /build/pylint-2.15.8/.pybuild/cpython3_3.10/build; python3.10 -m pytest -vvvv -k 'not test_pkginfo and not test_do_not_import_files_from_local_directory and not test_import_plugin_from_local_directory_if_pythonpath_cwd and not test_can_list_directo
ries_without_dunder_init and not test_fail_on_exit_code and not test__test_environ_pythonpath_no_arg'
============================= test session starts ==============================
platform linux -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0+repack -- /usr/bin/python3.10
cachedir: .pytest_cache
benchmark: 3.2.2 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /build/pylint-2.15.8, configfile: setup.cfg
plugins: benchmark-3.2.2
collecting ... collected 2056 items / 15 deselected / 2041 selected

tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_initialize PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_initialize_pickling PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_check_single_file_uninitialised PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_check_single_file_no_checkers PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_linter_with_unpickleable_plugins_is_pickleable PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallelFramework::test_worker_check_sequential_checker PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallel::test_sequential_checkers_work PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallel::test_invoke_single_job PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallel::test_compare_workers_to_single_proc[1-2-1] PASSED [  0%]
tests/test_check_parallel.py::TestCheckParallel::test_compare_workers_to_single_proc[1-2-2] PASSED [  0%]
...
```

this is running in debian unstable, with astroid/2.12.13

### Configuration

_No response_

### Command used

```shell
pytest command and switches reported above
```

### Pylint output

```shell
n/a
```

### Expected behavior

unittests to work

### Pylint version

```shell
pylint 2.15.8
astroid 2.12.13
python 3.11.1
```

### OS / Environment

Debian unstable

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by sandrotosi**
Fixes #7916

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -33,7 +33,7 @@ classifiers = [
 ]
 requires-python = ">=3.7.2"
 dependencies    = [
-    "dill>=0.2",
+    "dill>=0.3.6",
```
Comment (Pierre-Sassoulas) [inline pyproject.toml:36]:
```suggestion
    "dill>=0.2;python_version<'3.11'",
    "dill>=0.3.6;python_version>='3.11'",
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you @sandrotosi  ! The fix I did is based on https://github.com/PyCQA/pylint/pull/7838/files#r1031123153. Look like dill did not handle metadata correctly and we do need to be precise if we want to support 3.7 to 3.11.

**Review by Pierre-Sassoulas — DISMISSED** [review]

**Comment by Pierre-Sassoulas** [comment]
Seems like forcing the CI environment to be cache invalidated made a pre-existing issue appear for pypy 3.8 (brought by #7863 ?). I'll check that later on.

**Review by DanielNoord — APPROVED** [review]

**Review by DanielNoord — APPROVED** [review]

**Review by ssbarnea — APPROVED** [review]

**Comment by Pierre-Sassoulas** [comment]
Merging because the pypy 3.8 problem exists on main anyway.

**Comment by Pierre-Sassoulas** [comment]
Backported in https://github.com/PyCQA/pylint/commit/83668de8080a8adf2393dcda826100c6dba8267c

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Do we know why `main` is broken?

**Comment by Pierre-Sassoulas** [comment]
No idea, the error (only on pypy 3.8) is very strange and seem related to #7863 (?)

**Comment by DanielNoord** [comment]
Should we revert that commit then?

**Comment by Pierre-Sassoulas** [comment]
It's just an intuition based on the test fail message (so : maybe 😄 ). TBH I don't have a lot of respect for pypy as I never used it myself so just excluding the interpreter in conf would not bother me.

**Comment by GideonBear** [comment]
Second time my PR causes failures on *just* PyPy 3.8 :sob:

**Comment by Pierre-Sassoulas** [comment]
The bug is really strange btw:
```
@dataclass
class CustomDataClass: #  multiple-statements ??!
    my_var: list[int]

@dataclasses.dataclass
class CustomDataClass2: #  multiple-statements ??!
    my_var: list[int]

@dataclass()
class CustomDataClass3: #  multiple-statements ??!
    my_var: list[int]
```

**Comment by GideonBear** [comment]
@Pierre-Sassoulas
Should the discussion about this failure happen in a separate issue?
I reproduced it on my local machine, I'm now trying to debug it.

**Comment by Pierre-Sassoulas** [comment]
See #7945 for a hot fix, if we have a real fix then we can apply it instead.

**Pull request merged**